test/sql/cast/string_to_struct_cast.test

```diff
@@ -0,0 +1,463 @@
+----
+[["hello"], ["world"], ["!"]]
+[["Amazing"], ["text"]]
+[[Hello World!]]
+
+# ---------------------------------------------------
+statement  ok
+CREATE TABLE superNestedStrings (col1 VARCHAR)
+
+statement ok
+INSERT INTO superNestedStrings VALUES ('[[[[["hello"]]], [[["world"],["!"]]]]]'), ('[[[[["Amazing"]],[["text"]]]]]'), ('[[[[[Hello World!]]]]]');
+
+query I
+SELECT col1::VARCHAR[][][][][] FROM superNestedStrings;
+----
+[[[[["hello"]]], [[["world"], ["!"]]]]]
+[[[[["Amazing"]], [["text"]]]]]
+[[[[[Hello World!]]]]]
+
+# ---------------------------------------------------
+statement ok
+CREATE TABLE tbl (col1 VARCHAR);
+
+statement ok
+INSERT INTO tbl VALUES ('[1,2,2]'), ('[345,67865,44,4]'), ('[5,6,7]');
+
+query I
+SELECT col1::INT[] FROM tbl;
+----
+[1, 2, 2]
+[345, 67865, 44, 4]
+[5, 6, 7]
+
+
+# ---------------------------------------------------
+statement ok
+CREATE TABLE doubleNested (col1 VARCHAR);
+
+statement ok
+INSERT INTO doubleNested VALUES ('[[1,2],[2]]'), ('[[345],[67865,44,4]]'), ('[[5],[6,7]]');
+
+query I
+SELECT col1::INT[][] FROM doubleNested;
+----
+[[1, 2], [2]]
+[[345], [67865, 44, 4]]
+[[5], [6, 7]]
+
+
+# ---------------------------------------------------
+statement  ok
+CREATE TABLE tripleNested (col1 VARCHAR)
+
+statement ok
+INSERT INTO tripleNested VALUES ('[[[1,2],[3]]]'), ('[[[4]]]');
+
+query I
+SELECT col1::INT[][][] FROM tripleNested;
+----
+[[[1, 2], [3]]]
+[[[4]]]
+
+# ---------------------------------------------------
+statement  ok
+CREATE TABLE crazyNested (col1 VARCHAR)
+
+statement ok
+INSERT INTO crazyNested VALUES ('[[[[[[1]],[[3,4,9]]],[[[0,1]]]]]]'), ('[[[[[[4]]]]]]');
+
+query I
+SELECT col1::INT[][][][][][] FROM crazyNested;
+----
+[[[[[[1]], [[3, 4, 9]]], [[[0, 1]]]]]]
+[[[[[[4]]]]]]
+
+
+#               Test WHERE clause
+# ---------------------------------------------------
+query I
+SELECT * FROM tbl WHERE cast(col1 as int[]) = [1, 2, 2];
+----
+[1,2,2]
+
+query I
+SELECT col1 FROM tbl WHERE LEN(cast(col1 as int[])) < 4;
+----
+[1,2,2]
+[5,6,7]
+
+query I
+SELECT cast(col1 as int[]) FROM tbl WHERE LEN(cast(col1 as int[])) < 4;
+----
+[1, 2, 2]
+[5, 6, 7]
+
+
+#               All types:
+# ---------------------------------------------------
+
+query I
+SELECT CAST('[[True,False], [true], [FALSE], [false, true]]' AS BOOL[][]);
+----
+[[true, false], [true], [false], [false, true]]
+
+foreach type TINYINT SMALLINT INTEGER BIGINT HUGEINT UTINYINT USMALLINT UINTEGER UBIGINT 
+
+query I
+SELECT CAST('[1,2,3]' AS ${type}[]);
+----
+[1, 2, 3]
+
+endloop
+
+foreach type DOUBLE[] FLOAT[]
+
+query I
+SELECT CAST('[[1,2,3], [3.333, 2.2, 1.1], [0]]' AS ${type}[]);
+----
+[[1.0, 2.0, 3.0], [3.333, 2.2, 1.1], [0.0]]
+
+endloop
+
+statement ok
+CREATE TYPE mood AS ENUM('ok', 'sad', 'happy');
+
+query I
+SELECT '[ok,happy]'::MOOD[]
+----
+[ok, happy]
+
+statement error
+SELECT '[nothappy]'::MOOD[]
+
+query I
+SELECT CAST(list_value(date)::VARCHAR AS DATE[]) FROM test_all_types();
+----
```

Comment:
Empty lists should be castable to any other list (including nested lists):

```sql
SELECT '[]'::INT[];
SELECT '[]'::INT[][];
SELECT '[]'::INT[][][][];

```

This should also work with nested lists:

```sql
SELECT '[[1, 2, 3], [], NULL, [NULL], [4, NULL]]'::INT[][]
```